README.rst
mcu-uuid-syslog |Build Status|
==============================

Description
-----------

Microcontroller syslog service

Purpose
-------

Provides a log handler that sends messages to a syslog server (using
the `RFC 5424 protocol <https://tools.ietf.org/html/rfc5424>`_).

Documentation
-------------

`Read the documentation <https://mcu-uuid-syslog.readthedocs.io/>`_
generated from the docs_ directory.

.. _docs: docs/

.. |Build Status| image:: https://jenkins.uuid.uk/buildStatus/icon?job=mcu-uuid-syslog%2Fmain


docs/changelog.rst
Change log
==========

Unreleased_
-----------

2.2.2_ |--| 2022-11-07
----------------------

Fix handling of (invalid) facility values higher than 31.

Changed
~~~~~~~

* Use ``PSTR_ALIGN`` for flash strings.

Fixed
~~~~~

* Facility values higher than 31 (values higher than 23 are not allowed
  by the ``uuid::log::Facility`` enum) would cause the priority value to
  be out of range (more than 3 digits).

2.2.1_ |--| 2022-10-29
----------------------

Fix thread-safe mode and mark message handling.

Changed
~~~~~~~

* Limit the maximum number of log messages that can be transmitted in
  one loop iteration.

Fixed
~~~~~

* Crash when transmitting messages when operating in thread-safe mode.
* Last message time wasn't being updated so a flood of mark messages
  would occur when the uptime reached the mark interval.

2.2.0_ |--| 2022-10-26
----------------------

Be thread-safe (for log messages) where possible.

Added
~~~~~

* Indicate whether this version of the library is thread-safe or not
  (``UUID_SYSLOG_THREAD_SAFE`` and ``uuid::syslog::thread_safe``).

Changed
~~~~~~~

* Make the library thread-safe (for log messages only) when supported by the
  platform.

2.1.2_ |--| 2022-02-28
----------------------

Fix performance on the ESP32.

Fixed
~~~~~

* Include lwIP headers directly so that messages can be sent faster on the
  ESP32.

2.1.1_ |--| 2022-02-23
----------------------

Silence irrelevant compiler warnings when building on the ESP32.

Changed
~~~~~~~

* Silence compiler warnings/errors about ``%S`` in a format string.

2.1.0_ |--| 2022-01-27
----------------------

More visibility of the log message queue and control over the rate at which
messages are dispatched.

Added
~~~~~

* Function to get the current size of the log message queue.
* Symbols to control the UDP message sending delay:

  * ``UUID_SYSLOG_UDP_BASE_MESSAGE_DELAY``
  * ``UUID_SYSLOG_UDP_IPV4_ARP_MESSAGE_DELAY``
  * ``UUID_SYSLOG_UDP_IPV6_NDP_MESSAGE_DELAY``

  It is inadvisable to change these because UDP packets may be queued and then
  discarded by the platform when the queue limit is reached.

Changed
~~~~~~~

* Relax IPv6 scope checking when waiting for a local address to allow
  either global or local unicast addresses to be used with any global
  or local unicast destination. Previously they had to be the same type.

2.0.6_ |--| 2021-04-18
----------------------

Upgrade to PlatformIO 5.

Changed
~~~~~~~

* Use PlatformIO 5 dependency specification.

2.0.5_ |--| 2021-01-17
----------------------

Upgrade to the latest version of the logging library for static
initialization and deinitialization fixes.

Changed
~~~~~~~

* Don't unregister handler explicitly in the destructor, this is now
  handled by the logging library.

2.0.4_ |--| 2019-09-22
----------------------

Fix log message transmit retries.

Fixed
~~~~~

* Log messages that failed to be sent are not left on the queue
  correctly and may cause a crash when they are retried.
* Add memory barrier around checks for log queue overflow.

2.0.3_ |--| 2019-09-21
----------------------

Feature detection fixes.

Fixed
~~~~~

* Use ``gettimeofday()`` on the ESP32.
* Don't redefine ``UUID_SYSLOG_ARP_CHECK``.

2.0.2_ |--| 2019-09-20
----------------------

Support IPv6 addresses.

Fixed
~~~~~

* Use move constructors on rvalues.
* Support for IPv6 addresses.

2.0.1_ |--| 2019-09-05
----------------------

Bug fix for trace level messages.

Fixed
~~~~~

* Use debug level for trace level messages.

2.0.0_ |--| 2019-09-03
----------------------

Additional features and API changes.

Added
~~~~~

* Functions to get all of the configuration parameters.
* Support for sending a ``-- MARK --`` message when there is no
  activity for a configurable period of time.

Changed
~~~~~~~

* Add log level to the message text.
* Lower the log level of log level change messages to ``INFO``.
* Rename ``set_host()`` to ``destination()``.
* Remove ``get_`` and ``set_`` from function names.

Fixed
~~~~~

* Function to set the maximum number of log messages is no longer
  limited to 0 or 1.

1.0.0_ |--| 2019-09-01
----------------------

First stable release.

Added
~~~~~

* Buffer messages during startup until configuration is provided.
* Automatic use of system time if it is available.
* Wait for the network to be available before transmitting.
* Explicit ARP check for the destination host on ESP8266 and ESP32.
* Rate limiting of output packets to avoid dropped messages.
* Configurable destination host and log level.
* Configurable local hostname.
* Configurable queue size.

.. |--| unicode:: U+2013 .. EN DASH

.. _Unreleased: https://github.com/nomis/mcu-uuid-syslog/compare/2.2.2...HEAD
.. _2.2.2: https://github.com/nomis/mcu-uuid-syslog/compare/2.2.1...2.2.2
.. _2.2.1: https://github.com/nomis/mcu-uuid-syslog/compare/2.2.0...2.2.1
.. _2.2.0: https://github.com/nomis/mcu-uuid-syslog/compare/2.1.2...2.2.0
.. _2.1.2: https://github.com/nomis/mcu-uuid-syslog/compare/2.1.1...2.1.2
.. _2.1.1: https://github.com/nomis/mcu-uuid-syslog/compare/2.1.0...2.1.1
.. _2.1.0: https://github.com/nomis/mcu-uuid-syslog/compare/2.0.6...2.1.0
.. _2.0.6: https://github.com/nomis/mcu-uuid-syslog/compare/2.0.5...2.0.6
.. _2.0.5: https://github.com/nomis/mcu-uuid-syslog/compare/2.0.4...2.0.5
.. _2.0.4: https://github.com/nomis/mcu-uuid-syslog/compare/2.0.3...2.0.4
.. _2.0.3: https://github.com/nomis/mcu-uuid-syslog/compare/2.0.2...2.0.3
.. _2.0.2: https://github.com/nomis/mcu-uuid-syslog/compare/2.0.1...2.0.2
.. _2.0.1: https://github.com/nomis/mcu-uuid-syslog/compare/2.0.0...2.0.1
.. _2.0.0: https://github.com/nomis/mcu-uuid-syslog/compare/1.0.0...2.0.0
.. _1.0.0: https://github.com/nomis/mcu-uuid-syslog/commits/1.0.0


src/uuid/syslog.h
/*
 * uuid-syslog - Microcontroller syslog service
 * Copyright 2019,2021-2022  Simon Arlott
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef UUID_SYSLOG_H_
#define UUID_SYSLOG_H_

#include <Arduino.h>
#ifdef ARDUINO_ARCH_ESP8266
# include <ESP8266WiFi.h>
#else
# include <WiFi.h>
#endif
#include <WiFiUdp.h>
#include <time.h>

#include <list>
#include <memory>
#include <string>

#include <uuid/log.h>

#ifndef UUID_LOG_THREAD_SAFE
# define UUID_LOG_THREAD_SAFE 0
#endif

#ifndef UUID_COMMON_STD_MUTEX_AVAILABLE
# define UUID_COMMON_STD_MUTEX_AVAILABLE 0
#endif

#if defined(DOXYGEN) || UUID_COMMON_STD_MUTEX_AVAILABLE
# define UUID_SYSLOG_THREAD_SAFE 1
#else
# define UUID_SYSLOG_THREAD_SAFE 0
#endif

#if UUID_LOG_THREAD_SAFE
# include <mutex>
#endif

namespace uuid {

/**
 * Syslog service.
 *
 * - <a href="https://github.com/nomis/mcu-uuid-syslog/">Git Repository</a>
 * - <a href="https://mcu-uuid-syslog.readthedocs.io/">Documentation</a>
 */
namespace syslog {

/**
 * Thread-safe status of the library.
 *
 * @since 2.2.0
 */
#if UUID_COMMON_THREAD_SAFE && UUID_LOG_THREAD_SAFE && UUID_SYSLOG_THREAD_SAFE
static constexpr bool thread_safe = true;
#else
static constexpr bool thread_safe = false;
#endif

/**
 * Log handler for sending messages to a syslog server.
 *
 * @since 1.0.0
 */
class SyslogService: public uuid::log::Handler {
public:
	static constexpr size_t MAX_LOG_MESSAGES = 50; /*!< Maximum number of log messages to buffer before they are output. @since 1.0.0 */
	static constexpr uint16_t DEFAULT_PORT = 514; /*!< Default UDP port to send messages to. @since 1.0.0 */

	/**
	 * Create a new syslog service log handler.
	 *
	 * @since 1.0.0
	 */
	SyslogService() = default;
	~SyslogService() = default;

	/**
	 * Register the log handler with the logging framework.
	 *
	 * @since 1.0.0
	 */
	void start();

	/**
	 * Get the current log level.
	 *
	 * This only affects newly received log messages, not messages that
	 * have already been queued.
	 *
	 * @return The current log level.
	 * @since 2.0.0
	 */
	uuid::log::Level log_level() const;
	/**
	 * Set the current log level.
	 *
	 * Unless this is the first time the log level is being set, this
	 * only affects newly received log messages, not messages that have
	 * already been queued.
	 *
	 * @param[in] level Minimum log level that will be sent to the
	 *                  syslog server.
	 * @since 2.0.0
	 */
	void log_level(uuid::log::Level level);

	/**
	 * Get the maximum number of queued log messages.
	 *
	 * @return The maximum number of queued log messages.
	 * @since 2.0.0
	 */
	size_t maximum_log_messages() const;
	/**
	 * Set the maximum number of queued log messages.
	 *
	 * Defaults to SyslogService::MAX_LOG_MESSAGES.
	 *
	 * @since 2.0.0
	 */
	void maximum_log_messages(size_t count);

	/**
	 * Get the current number of queued log messages.
	 *
	 * @return The current number of queued log messages.
	 * @since 2.1.0
	 */
	size_t current_log_messages() const;

	/**
	 * Get the server to send messages to.
	 *
	 * @since 2.0.0
	 * @return IP address and UDP port of the syslog server.
	 */
	std::pair<IPAddress,uint16_t> destination() const;
	/**
	 * Set the server to send messages to.
	 *
	 * To disable sending messages, set the host to `0.0.0.0` and the
	 * log level to uuid::log::Level::OFF (otherwise they will be
	 * queued but not sent).
	 *
	 * @param[in] host IP address of the syslog server.
	 * @param[in] port UDP port to send messages to.
	 * @since 2.0.0
	 */
	void destination(IPAddress host, uint16_t port = DEFAULT_PORT);

	/**
	 * Get local hostname.
	 *
	 * @since 2.0.0
	 * @return Hostname of this device.
	 */
	std::string hostname() const;
	/**
	 * Set local hostname.
	 *
	 * @param[in] hostname Hostname of this device.
	 * @since 2.0.0
	 */
	void hostname(std::string hostname);

	/**
	 * Get mark interval.
	 *
	 * @since 2.0.0
	 * @return Mark interval in seconds (0 = disable).
	 */
	unsigned long mark_interval() const;
	/**
	 * Set mark interval.
	 *
	 * When no messages have been sent for this period of time, a
	 * `-- MARK --` message will be generated automatically.
	 *
	 * @param[in] interval Mark interval in seconds (0 = disable).
	 * @since 2.0.0
	 */
	void mark_interval(unsigned long interval);

	/**
	 * Dispatch queued log messages.
	 *
	 * @since 1.0.0
	 */
	void loop();

	/**
	 * Add a new log message.
	 *
	 * This will be put in a queue for output at the next loop()
	 * process. The queue has a maximum size of
	 * get_maximum_log_messages() and will discard the oldest message
	 * first.
	 *
	 * @param[in] message New log message, shared by all handlers.
	 * @since 1.0.0
	 */
	virtual void operator<<(std::shared_ptr<uuid::log::Message> message);

private:
	/**
	 * Log message that has been queued.
	 *
	 * Contains an identifier sequence to indicate when log messages
	 * could not be output because the queue discarded one or more
	 * messages.
	 *
	 * @since 1.0.0
	 */
	class QueuedLogMessage {
	public:
		/**
		 * Create a queued log message.
		 *
		 * @param[in] id Identifier to use for the log message on the queue.
		 * @param[in] content Log message content.
		 * @since 1.0.0
		 */
		QueuedLogMessage(unsigned long id, std::shared_ptr<uuid::log::Message> &&content);
		~QueuedLogMessage() = default;

		unsigned long id_; /*!< Sequential identifier for this log message. @since 1.0.0 */
		struct timeval time_; /*!< Time message was received. @since 1.0.0 */
		std::shared_ptr<const uuid::log::Message> content_; /*!< Log message content. @since 1.0.0 */

	private:
		static bool time_good_; /*!< System time appears to be valid. @since 1.0.0 */
	};

	/**
	 * Add a new log message.
	 *
	 * This will be put in a queue for output at the next loop()
	 * process. The queue has a maximum size of
	 * get_maximum_log_messages() and will discard the oldest message
	 * first.
	 *
	 * @param[in] message New log message, shared by all handlers.
	 * @since 2.2.0
	 */
	void add_message(std::shared_ptr<uuid::log::Message> &message);

	/**
	 * Remove messages that were queued before the log level was set.
	 *
	 * @param[in] level New log level
	 * @since 1.0.0
	 */
	void remove_queued_messages(uuid::log::Level level);

	/**
	 * Check if it is possible to transmit to the server.
	 *
	 * @return True if it is safe to transmit a message to the server,
	 *         otherwise false.
	 * @since 1.0.0
	 */
	bool can_transmit();

	/**
	 * Attempt to transmit one message to the server.
	 *
	 * @param[in] message Log message to be sent.
	 * @return True if the message was successfully set, otherwise
	 *         false.
	 * @since 1.0.0
	 */
	bool transmit(const QueuedLogMessage &message);

	static uuid::log::Logger logger_; /*!< uuid::log::Logger instance for syslog services. @since 1.0.0 */

	bool started_ = false; /*!< Flag to indicate that messages have started being transmitted. @since 1.0.0 */
	bool level_set_ = false; /*!< Flag to indicate that the log level has been set at least once. @since 2.2.0 */
	WiFiUDP udp_; /*!< UDP client. @since 1.0.0 */
	IPAddress host_; /*!< Host to send messages to. @since 1.0.0 */
	uint16_t port_ = DEFAULT_PORT; /*!< Port to send messages to. @since 1.0.0 */
	uint64_t last_transmit_ = 0; /*!< Last transmit time. @since 1.0.0 */
	std::string hostname_{'-'}; /*!< Local hostname. @since 1.0.0 */
#if UUID_SYSLOG_THREAD_SAFE
	mutable std::mutex mutex_; /*!< Mutex for queued log messages. @since 2.2.0 */
#endif
	size_t maximum_log_messages_ = MAX_LOG_MESSAGES; /*!< Maximum number of log messages to buffer before they are output. @since 1.0.0 */
	unsigned long log_message_id_ = 0; /*!< The next identifier to use for queued log messages. @since 1.0.0 */
	std::list<QueuedLogMessage> log_messages_; /*!< Queued log messages, in the order they were received. @since 1.0.0 */
	uint64_t mark_interval_ = 0; /*!< Mark interval in milliseconds. @since 2.0.0 */
	uint64_t last_message_ = 0; /*!< Last message/mark time. @since 2.0.0 */
};

} // namespace syslog

} // namespace uuid

#endif


examples/Example.cpp
#include <Arduino.h>
#ifdef ARDUINO_ARCH_ESP8266
# include <ESP8266WiFi.h>
#else
# include <WiFi.h>
#endif
#include <uuid/common.h>
#include <uuid/log.h>
#include <uuid/syslog.h>

static uuid::syslog::SyslogService syslog;

void setup() {
	static uuid::log::Logger logger{F("setup")};

	syslog.start();
	/* Retrieve settings after calling start()
	 * in case the settings read process logs
	 * some messages.
	 */
	syslog.hostname("example");
	syslog.log_level(uuid::log::DEBUG);
	syslog.mark_interval(3600);
	syslog.destination(IPAddress(192, 0, 2, 1));

	WiFi.persistent(false);
	WiFi.mode(WIFI_STA);
	WiFi.begin("SSID", "password");

	Serial.begin(115200);

	logger.info(F("Application started"));
}

void loop() {
	static uuid::log::Logger logger{F("loop")};
	static unsigned int i = 0;

	uuid::loop();
	syslog.loop();

	logger.debug(F("Hello %u World!"), i++);

	delay(1000);
}


src/syslog.cpp
/*
 * uuid-syslog - Microcontroller syslog service
 * Copyright 2019,2021-2022  Simon Arlott
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "uuid/syslog.h"

#include <Arduino.h>
#ifdef ARDUINO_ARCH_ESP8266
# include <ESP8266WiFi.h>
#else
# include <WiFi.h>
#endif
#include <WiFiUdp.h>

#if defined __has_include
# if __has_include (<lwip/init.h>)
#  include <lwip/init.h>
# endif
#endif

#ifndef UUID_SYSLOG_HAVE_GETTIMEOFDAY
# if defined(ARDUINO_ARCH_ESP8266) || defined(ARDUINO_ARCH_ESP32)
// time() does not return UTC on the ESP8266: https://github.com/esp8266/Arduino/issues/4637
#  define UUID_SYSLOG_HAVE_GETTIMEOFDAY 1
# endif
#endif

#ifndef UUID_SYSLOG_HAVE_GETTIMEOFDAY
# define UUID_SYSLOG_HAVE_GETTIMEOFDAY 0
#endif

#ifndef UUID_SYSLOG_HAVE_IPADDRESS_TYPE
# if defined(ARDUINO_ARCH_ESP8266)
#  define UUID_SYSLOG_HAVE_IPADDRESS_TYPE 1
# endif
#endif

#ifndef UUID_SYSLOG_HAVE_IPADDRESS_TYPE
# define UUID_SYSLOG_HAVE_IPADDRESS_TYPE 0
#endif

#ifndef UUID_SYSLOG_ARP_CHECK
# if defined(LWIP_VERSION_MAJOR) && defined(LWIP_IPV4) && LWIP_VERSION_MAJOR >= 2 && LWIP_IPV4
#  define UUID_SYSLOG_ARP_CHECK 1
# endif
#endif

#ifndef UUID_SYSLOG_ARP_CHECK
# define UUID_SYSLOG_ARP_CHECK 0
#endif

#ifndef UUID_SYSLOG_NDP_CHECK
# if defined(LWIP_VERSION_MAJOR) && defined(LWIP_IPV6) && LWIP_VERSION_MAJOR >= 2 && LWIP_IPV6
#  define UUID_SYSLOG_NDP_CHECK 1
# endif
#endif

#ifndef UUID_SYSLOG_NDP_CHECK
# define UUID_SYSLOG_NDP_CHECK 0
#endif

#if UUID_SYSLOG_ARP_CHECK or UUID_SYSLOG_NDP_CHECK
# include <lwip/netif.h>
#endif
#if UUID_SYSLOG_ARP_CHECK
# include <lwip/ip4_addr.h>
# include <lwip/etharp.h>
#endif
#if UUID_SYSLOG_NDP_CHECK
# include <lwip/ip6_addr.h>
# include <lwip/nd6.h>
#endif

#include <algorithm>
#include <list>
#include <memory>
#if UUID_SYSLOG_THREAD_SAFE
# include <mutex>
#endif
#include <string>

#include <uuid/common.h>
#include <uuid/log.h>

#ifndef UUID_SYSLOG_UDP_BASE_MESSAGE_DELAY
# define UUID_SYSLOG_UDP_BASE_MESSAGE_DELAY 100
#endif

#ifndef UUID_SYSLOG_UDP_IPV4_ARP_MESSAGE_DELAY
# define UUID_SYSLOG_UDP_IPV4_ARP_MESSAGE_DELAY 10
#endif

#ifndef UUID_SYSLOG_UDP_IPV6_NDP_MESSAGE_DELAY
# define UUID_SYSLOG_UDP_IPV6_NDP_MESSAGE_DELAY 10
#endif

#ifndef PSTR_ALIGN
# define PSTR_ALIGN 4
#endif

static const char __pstr__logger_name[] __attribute__((__aligned__(PSTR_ALIGN))) PROGMEM = "syslog";

namespace uuid {

namespace syslog {

uuid::log::Logger SyslogService::logger_{FPSTR(__pstr__logger_name), uuid::log::Facility::SYSLOG};
bool SyslogService::QueuedLogMessage::time_good_ = false;

void SyslogService::start() {
	uuid::log::Logger::register_handler(this, uuid::log::Level::ALL);
}

uuid::log::Level SyslogService::log_level() const {
	return uuid::log::Logger::get_log_level(this);
}

void SyslogService::remove_queued_messages(uuid::log::Level level) {
#if UUID_SYSLOG_THREAD_SAFE
	std::lock_guard<std::mutex> lock{mutex_};
#endif
	unsigned long offset = 0;

	for (auto it = log_messages_.begin(); it != log_messages_.end(); ) {
		if (it->content_->level > level) {
			offset++;
			it = log_messages_.erase(it);
		} else {
			it->id_ -= offset;
			it++;
		}
	}

	log_message_id_ -= offset;
}

void SyslogService::log_level(uuid::log::Level level) {
	if (!started_) {
		remove_queued_messages(level);
	}

	bool level_changed = !level_set_ || (level != log_level());
	level_set_ = true;

	if (level_changed && level < uuid::log::Level::NOTICE) {
		logger_.info(F("Log level set to %S"), uuid::log::format_level_uppercase(level));
	}
	uuid::log::Logger::register_handler(this, level);
	if (level_changed && level >= uuid::log::Level::NOTICE) {
		logger_.info(F("Log level set to %S"), uuid::log::format_level_uppercase(level));
	}
}

size_t SyslogService::maximum_log_messages() const {
#if UUID_SYSLOG_THREAD_SAFE
	std::lock_guard<std::mutex> lock{mutex_};
#endif

	return maximum_log_messages_;
}

void SyslogService::maximum_log_messages(size_t count) {
#if UUID_SYSLOG_THREAD_SAFE
	std::lock_guard<std::mutex> lock{mutex_};
#endif

	maximum_log_messages_ = std::max((size_t)1, count);

	while (log_messages_.size() > maximum_log_messages_) {
		log_messages_.pop_front();
	}
}

size_t SyslogService::current_log_messages() const {
#if UUID_SYSLOG_THREAD_SAFE
	std::lock_guard<std::mutex> lock{mutex_};
#endif

	return log_messages_.size();
}

std::pair<IPAddress,uint16_t> SyslogService::destination() const {
	return std::make_pair(host_, port_);
}

void SyslogService::destination(IPAddress host, uint16_t port) {
	host_ = host;
	port_ = port;

	if ((uint32_t)host_ == (uint32_t)0) {
		started_ = false;
		remove_queued_messages(log_level());
	}
}

std::string SyslogService::hostname() const {
	return hostname_;
}

void SyslogService::hostname(std::string hostname) {
	if (hostname.empty() || hostname.find(' ') != std::string::npos) {
		hostname_ = '-';
	} else {
		hostname_ = std::move(hostname);
	}
}

unsigned long SyslogService::mark_interval() const {
	return mark_interval_ / 1000;
}

void SyslogService::mark_interval(unsigned long interval) {
	mark_interval_ = (uint64_t)interval * 1000;
}

SyslogService::QueuedLogMessage::QueuedLogMessage(unsigned long id, std::shared_ptr<uuid::log::Message> &&content)
		: id_(id), content_(std::move(content)) {
	if (time_good_ || WiFi.status() == WL_CONNECTED) {
#if UUID_SYSLOG_HAVE_GETTIMEOFDAY
		if (gettimeofday(&time_, nullptr) != 0) {
			time_.tv_sec = (time_t)-1;
		}
#else
		time_.tv_sec = time(nullptr);
		time_.tv_usec = 0;
#endif

		if (time_.tv_sec >= 0 && time_.tv_sec < 18140 * 86400) {
			time_.tv_sec = (time_t)-1;
		}

		if (time_.tv_sec != (time_t)-1) {
			time_good_ = true;
		}
	} else {
		time_.tv_sec = (time_t)-1;
	}
}

/* Mutex already locked by caller. */
void SyslogService::add_message(std::shared_ptr<uuid::log::Message> &message) {
	if (log_messages_.size() >= maximum_log_messages_) {
		log_messages_.pop_front();
	}

	log_messages_.emplace_back(log_message_id_++, std::move(message));
}

void SyslogService::operator<<(std::shared_ptr<uuid::log::Message> message) {
#if UUID_SYSLOG_THREAD_SAFE
	std::lock_guard<std::mutex> lock{mutex_};
#endif

	add_message(message);
}

void SyslogService::loop() {
#if UUID_SYSLOG_THREAD_SAFE
	std::unique_lock<std::mutex> lock{mutex_};
#endif
	size_t count = std::max((size_t)1, MAX_LOG_MESSAGES);

	while (!log_messages_.empty()) {
#if UUID_SYSLOG_THREAD_SAFE
		lock.unlock();
#endif

		if (!can_transmit())
			return;

#if UUID_SYSLOG_THREAD_SAFE
		lock.lock();

		if (log_messages_.empty())
			break;
#endif

		auto message = log_messages_.front();

		started_ = true;

#if UUID_SYSLOG_THREAD_SAFE
		lock.unlock();
#endif

		auto ok = transmit(message);
		if (ok) {
#if UUID_SYSLOG_THREAD_SAFE
			lock.lock();
#endif

			last_message_ = last_transmit_;
			if (!log_messages_.empty()
					&& log_messages_.front().content_ == message.content_) {
				log_messages_.pop_front();
			}

#if UUID_SYSLOG_THREAD_SAFE
			lock.unlock();
#endif
		}

		::yield();

#if UUID_SYSLOG_THREAD_SAFE
		lock.lock();
#endif

		if (!ok) {
			break;
		}

		count--;
		if (count == 0) {
			break;
		}
	}

	if (started_ && mark_interval_ != 0 && log_messages_.empty()) {
		if (uuid::get_uptime_ms() - last_message_ >= mark_interval_) {
			// This is generated manually because the log level may not
			// be high enough to receive INFO messages.
			auto message = std::make_shared<uuid::log::Message>(uuid::get_uptime_ms(),
					uuid::log::Level::INFO,
					uuid::log::Facility::SYSLOG,
					reinterpret_cast<const __FlashStringHelper *>(__pstr__logger_name),
					uuid::read_flash_string(F("-- MARK --")));
			add_message(message);
		}
	}
}

bool SyslogService::can_transmit() {
#if UUID_SYSLOG_HAVE_IPADDRESS_TYPE
	if (host_.isV4() && (uint32_t)host_ == (uint32_t)0) {
		return false;
	}
#else
	if ((uint32_t)host_ == (uint32_t)0) {
		return false;
	}
#endif

	if (WiFi.status() != WL_CONNECTED) {
		return false;
	}

	const uint64_t now = uuid::get_uptime_ms();
	uint64_t message_delay = UUID_SYSLOG_UDP_BASE_MESSAGE_DELAY;

#if UUID_SYSLOG_ARP_CHECK
# if UUID_SYSLOG_HAVE_IPADDRESS_TYPE
	if (host_.isV4())
# endif
	{
		message_delay = UUID_SYSLOG_UDP_IPV4_ARP_MESSAGE_DELAY;
	}
#endif
#if UUID_SYSLOG_NDP_CHECK && UUID_SYSLOG_HAVE_IPADDRESS_TYPE
	if (host_.isV6()) {
		message_delay = UUID_SYSLOG_UDP_IPV6_NDP_MESSAGE_DELAY;
	}
#endif

	if (started_ && (now < last_transmit_ || now - last_transmit_ < message_delay)) {
		return false;
	}

#if UUID_SYSLOG_ARP_CHECK
# if UUID_SYSLOG_HAVE_IPADDRESS_TYPE
	if (host_.isV4())
# endif
	{
		ip4_addr_t ipaddr;

		ip4_addr_set_u32(&ipaddr, (uint32_t)host_);

		if (!ip4_addr_isloopback(&ipaddr)
				&& !ip4_addr_ismulticast(&ipaddr)
				&& !ip4_addr_isbroadcast(&ipaddr, netif_default)) {
			struct eth_addr *eth_ret = nullptr;
			const ip4_addr_t *ip_ret = nullptr;

			if (!ip4_addr_netcmp(&ipaddr, netif_ip4_addr(netif_default), netif_ip4_netmask(netif_default))) {
				// Replace addresses outside the network with the gateway address
				const ip4_addr_t *gw_addr = netif_ip4_gw(netif_default);

				if (gw_addr != nullptr) {
					ipaddr = *gw_addr;
				}
			}

			if (etharp_find_addr(netif_default, &ipaddr, &eth_ret, &ip_ret) == -1) {
				etharp_query(netif_default, &ipaddr, NULL);
				// Avoid querying lwIP again for 1 second
				last_transmit_ = uuid::get_uptime_ms() + (uint64_t)1000 - message_delay;

				return false;
			}
		}
	}
#endif

#if UUID_SYSLOG_NDP_CHECK && UUID_SYSLOG_HAVE_IPADDRESS_TYPE
	if (host_.isV6()) {
		ip6_addr_t ip6addr;

		IP6_ADDR(&ip6addr, host_.raw6()[0], host_.raw6()[1], host_.raw6()[2], host_.raw6()[3]);
		ip6_addr_assign_zone(&ip6addr, IP6_UNICAST, netif_default);

		if (!ip6_addr_isloopback(&ip6addr)
				&& !ip6_addr_ismulticast(&ip6addr)) {
			// Don't send to a scoped address until we have a valid address of the same type
			bool have_address = false;
			const u8_t *hwaddr = nullptr;

			for (size_t i = 0; i < LWIP_IPV6_NUM_ADDRESSES; i++) {
				if (ip6_addr_isvalid(netif_ip6_addr_state(netif_default, i))) {
					if (ip6_addr_islinklocal(&ip6addr)) {
						if (ip6_addr_islinklocal(netif_ip6_addr(netif_default, i))) {
							have_address = true;
							break;
						}
					} else if (ip6_addr_isglobal(&ip6addr)
							|| ip6_addr_isuniquelocal(&ip6addr)
							|| ip6_addr_issitelocal(&ip6addr)) {
						if (ip6_addr_isglobal(netif_ip6_addr(netif_default, i))
								|| ip6_addr_isuniquelocal(netif_ip6_addr(netif_default, i))
								|| ip6_addr_issitelocal(netif_ip6_addr(netif_default, i))) {
							have_address = true;
							break;
						}
					} else {
						have_address = true;
						break;
					}
				}
			}

			if (!have_address) {
				// Avoid checking lwIP again for 1 second
				last_transmit_ = uuid::get_uptime_ms() + (uint64_t)1000 - message_delay;

				return false;
			} else if (nd6_get_next_hop_addr_or_queue(netif_default, NULL, &ip6addr, &hwaddr) != ERR_OK) {
				// Avoid querying lwIP again for 1 second
				last_transmit_ = uuid::get_uptime_ms() + (uint64_t)1000 - message_delay;

				return false;
			}
		}
	}
#endif

	return true;
}

bool SyslogService::transmit(const QueuedLogMessage &message) {
	struct tm tm;

	tm.tm_year = 0;
	if (message.time_.tv_sec != (time_t)-1) {
		gmtime_r(&message.time_.tv_sec, &tm);
	}

	if (udp_.beginPacket(host_, port_) != 1) {
		last_transmit_ = uuid::get_uptime_ms();
		return false;
	}
	/*
	 * The level is constrained to 0-7 by design in RFC 5424 because higher
	 * severity values would be a different severity in another facility. The
	 * TRACE level and all other invalid values (because of the conversion to
	 * unsigned) are converted to DEBUG.
	 *
	 * RFC 5424 requires that the facility MUST be 0-23. Values here are allowed
	 * to go up to 31 because there is no obvious candidate to convert them to
	 * and because the multiplication by a factor of 256 means that (with the
	 * cast back to uint8_t) higher values just overflow into other facilities
	 * (which is easier than doing it by modulo 24). The enum doesn't allow
	 * values that are out of range of the RFC.
	 *
	 * The maximum possible priority value does not exceed the requirement that
	 * the PRI part MUST be 3-5 characters.
	 */
	udp_.printf_P(PSTR("<%u>1 "), (uint8_t)(message.content_->facility * 8U)
		+ std::min(7U, (unsigned int)message.content_->level));
	if (tm.tm_year != 0) {
		udp_.printf_P(PSTR("%04u-%02u-%02uT%02u:%02u:%02u.%06luZ"),
				tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday,
				tm.tm_hour, tm.tm_min, tm.tm_sec, (unsigned long)message.time_.tv_usec);
	} else {
		udp_.print('-');
	}
	udp_.printf_P(PSTR(" %s - - - - \xEF\xBB\xBF"), hostname_.c_str());
	udp_.print(uuid::log::format_timestamp_ms(message.content_->uptime_ms, 3).c_str());
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wformat"
	udp_.printf_P(PSTR(" %c %lu: [%S] "), uuid::log::format_level_char(message.content_->level), message.id_, message.content_->name);
#pragma GCC diagnostic pop
	udp_.print(message.content_->text.c_str());
	bool ok = (udp_.endPacket() == 1);

	last_transmit_ = uuid::get_uptime_ms();
	return ok;
}

} // namespace syslog

} // namespace uuid
